pep-9999.rst

```diff
@@ -0,0 +1,173 @@
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 29-Jan-2022
+Python-Version: 3.11
+Post-History:
+
+
+Abstract
+========
+
+Though float and Decimal types can represent `signed zero`_, in many fields
+of mathematics negative zero is surprising or unwanted-- especially in the
+in the context of displaying an (often rounded) numerical result.  This PEP
+proposes an extension to the string format specification, allowing negative
+zero to be normalized to positive zero.
+
+.. _`signed zero`:  https://en.wikipedia.org/wiki/Signed_zero
+
+
+Background
+==========
+
+Here is negative zero::
+
+    >>> x = -0.
+    >>> x
+    -0.0
+
+When formatting a number, negative zero can result from rounding.  Assuming
+the user's intention is truly to discard precision, the distinction between
+negative and positive zero of the rounded result might be considered an
+unwanted artifact::
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{x: .1f})
+     0.0
+    -0.0
+     0.1
+
+There are various approaches to clearing the sign of a negative zero.  It
+can be achieved without a conditional by adding positive zero::
+
+    >>> x = -0.
+    >>> x + 0.
+    0.0
+
+To normalize negative zero when formatting, it would be necessary to perform
+a redundant (and error-prone) pre-rounding of the input::
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{round(x, 1) + 0.: .1f}')
+     0.0
+     0.0
+     0.1
+
+The behavior of number formatting with regard to negative zero is not
+unique to Python.  The first low-level implementations of ``printf()``
+and such likely wired the sign bit to ``-`` in the output without special
+consideration of the zero mantissa case.  In hindsight, it would have
+been nice to have a formatting option for treatment of negative zero--
+perhaps even coercing it to positive zero by default.  However, for better
+or worse, existing number formatting implementations inherit from the
+original specimens, passing the zero sign through without option.
+
+
+Proposal
+========
+
+An optional, literal ``z`` is added to the
+`Format Specification Mini-Language`_ following ``sign``::
+
+    [[fill]align][sign][z][#][0][width][grouping_option][.precision][type]
+
+where ``z`` is allowed only for float and Decimal types.  The new specifier
+is supported by both f-strings and ``str.format()``, but not %-formatting.
+
+Synopsis::
+
+    >>> x = -.00001
+    >>> f'{x:z.1f}'
+    '0.0'
+
+    >>> x = decimal.Decimal('-.00001')
+    >>> '{}'.format('+z.1f')
+    '+0.0'
+
+    >>> x = 0
+    >>> f'{x:z.1f}'  # not supported on int type
+    ValueError: Negative zero coercion (z) not allowed in int format specifier
+
+    >>> x = -.00001
+    >>> '%z.1f' % x  # not supported on %-formatting
+    ValueError: unsupported format character 'z' (0x7a) at index 1
+
+.. _`Format Specification Mini-Language`: https://docs.python.org/3/library/string.html#format-specification-mini-language
+
+
+Rationale
+=========
```

Comment:
I wanted a section to concisely state the specification (currently Proposal), and another section to explain the rationale of the proposal.  It seems strange to explain why something is before stating what the thing is.  For PEP's, is it that "Rationale" implies "rationale of the motivation" rather than "rationale of the specification"?

Are there perhaps some other sections I can employ to match my intention?  Or do I explain why the specification is as it is within the Specification section itself?

Reply:
Hmm, I see how that can be potentially a little perplexing, and keep in mind that the rules aren't "hard and fast".

PEP 1 (PEP-0001) describes it this way:

> The rationale fleshes out the specification by describing why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages.
>
> The rationale should provide evidence of consensus within the community and discuss important objections or concerns raised during discussion.

And PEP 12 (PEP-0012) and the template say,

> [Describe why particular design decisions were made.]

Based on that and my impression of what is generally done in successful PEPs (others may differ):

* The Rationale section leads in to the specification by outlining the prior art/how the problem has been solved in other languages, linking/summarizing the consensus of any community discussions leading up to your PEP, outlining the high-level design decisions you made, and justifying how they were arrived at. For a short PEP like this, it needn't be too long.
* The Specification section then describes the full normative spec in explicit detail, which implements the design decisions discussed in the Rationale; it can sometimes briefly mention in context why a choice was made, though that is usually best kept fairly minimal and with clear reason (which you did).
* The Rejected Ideas section enumerates potential alternatives you considered or others suggested, and why they were not adopted.

So, what I would suggest for your leading Rationale section is:

* Your `Similar work` section would be a good fit near the top of the rationale section, as well as further detail you might want to add about prior art in other languages
* If you were involved in any public discussions that helped lead to the PEP, you might want to link and briefly summarize them
* Then, you could give a very high-level summary (even a sentence or two) of your design: a new formatting specifier to control this behavior, off by default (or you could move that latter part and its paragraph to a rejected idea), which the first three paragraphs, and the example, of your current rationale section lay out the reasoning for that high-level design choice. 

Then, you have your specification section, as it is now.

Finally, you can make your last couple paragraphs separate subsections in the Rejected Ideas section (and maybe the opt-out too), since they each discuss why a particular alternative was not adopted, along with any you might add based on my suggestions above, and any others suggest.

Does that make sense?

Reply:
> The rationale fleshes out the specification by describing why particular design decisions were made. 

This seems to reinforce that rationale is secondary to the specification.  So I wonder why the "fleshing out" is prescribed to come before the specification itself.  Is it worth reconsidering that order?

But I'll revise the document to confirm to the current guidelines-- thank you for the suggestions.